homeassistant/components/ads/__init__.py

```diff
@@ -222,37 +255,56 @@ def add_device_notification(self, name, plc_datatype, callback):
     def _device_notification_callback(self, notification, name):
         """Handle device notifications."""
         contents = notification.contents
-
         hnotify = int(contents.hNotification)
         _LOGGER.debug("Received notification %d", hnotify)
 
-        # get dynamically sized data array
+        # Get dynamically sized data array
         data_size = contents.cbSampleSize
-        data = (ctypes.c_ubyte * data_size).from_address(
+        data_address = (
             ctypes.addressof(contents)
             + pyads.structs.SAdsNotificationHeader.data.offset
         )
+        data = (ctypes.c_ubyte * data_size).from_address(data_address)
 
-        try:
-            with self._lock:
-                notification_item = self._notification_items[hnotify]
-        except KeyError:
+        # Acquire notification item
+        with self._lock:
+            notification_item = self._notification_items.get(hnotify)
+
+        if not notification_item:
             _LOGGER.error("Unknown device notification handle: %d", hnotify)
             return
 
-        # Parse data to desired datatype
-        if notification_item.plc_datatype == pyads.PLCTYPE_BOOL:
-            value = bool(struct.unpack("<?", bytearray(data))[0])
```

Comment:
we used to `bool` the response, we don't do that now, is that what we expect?

Reply:
corrected.